Add instructions for creating Wagtail blocks (#53)
mozmeao/platform-docs: docs/cms/blocks.md

# CMS blocks

Most pages defined in the CMS have at least one free form section that accepts different types of content and UI elements. To understand how to build those, read Wagtail's docs on [how to use StreamField for mixed content](https://docs.wagtail.org/en/stable/topics/streamfield.html).

## Block design principles

The UI elements available for each page type are first defined by designers and implemented according to the options the team wants to give to editors to create new pages. When implementing a new block, consider the following principles:

> - The basic usage of a block should be straight forward and intuitive.

Reviewer: ```suggestion
- The basic usage of a block should be straightforward and intuitive.
```

> - Set default values whenever possible so that editors don't have to make choices for the basic usage.

Reviewer: non-blocking: I think this is a good principle, but might deserve its own subsection: Basic Usage or Setting Defaults (what criteria do you apply to decide "basic usage")

Author: What do you think of this?
> Set default values whenever possible so that editors can think primarily about content and don't have to make choices when using a block. Pick what looks to be the simplest implementation of the block and make it the default.

- Separate content from presentation. Optional parameters, such as layout and theme, should be grouped inside a block's "Settings"

> - Protect editors from making mistakes. We trust our editors to spot when a page doesn't look good, but they shouldn't have to.

Reviewer: **blocking**: needs more info on how we should do this (validation? help text?)

Author: What about this?
> Protect editors from making mistakes. We trust our editors to spot when a page doesn't look good, but they shouldn't have to. If a specific combination of settings results in a bad UI, make that impossible. Add help text explaining the intended usage of each field.

  - If a block shouldn't be placed in a certain position on the page, add a page level validation.

>   - If a certain combination of settings in a block doesn't make sense, raise a validation error.

Reviewer: **non-blocking, structure**: Validation feels worthy of a separate subsection as well, so it's easy to find at a glance

>   - If a field is optional, make sure that the template doesn't render it, adding white space.

Reviewer: **non-blocking**: some of these points might also be useful to put in the Wagtail checklist section: https://mozmeao.github.io/platform-docs/code-review/checklists/backend/#wagtail-cms
I believe AGENTS.md points to these checklists for reviews

Author: Once this is merged, we should add this file to AGENTS.md too.

  - However blocks are combined in a page, they should always output semantic HTML. Headings hierarchy must be consistent, for example.
- Smaller UI elements (atoms and molecules) should be generic enough to be used inside other blocks with little or no extra configuration.

> - We frequently need to decide if a new UI element is a variation of an existing block or whole new one. To make that decision, consider these factors:

Reviewer: **non-blocking, structure**: love this! Would also pull it out to a subsection for easier scanning

  - The intended usage of the block. Does it make sense only in a specific page or could it be placed in any page?
  - Will the new settings addition to the existing block make it confusing for editors?
  - If implemented as a new block, will it be clear to editors which is which and why they should pick one or the other?
- Don't be afraid to refactor blocks if it means improving usability - data migrations exist for this. (See the section about [migrating blocks](#block-data-migrations) below.)

## Blocks hierarchy

Smaller UI elements are defined as blocks that are used inside other blocks. Those will rarely be used directly in pages. Some examples are:

- Various types of buttons
- Icons
- Headings
- Lists
- Images

Section-level blocks combine smaller blocks into more complex structures. They usually fill the page width. Some examples are:

- Intro
- Banner
- Media + Content
- Cards list

These are more opinionated UI pieces with responsive designs. They're usually the direct children of a page.

## Headings and buttons

Those are blocks that require some special templating logic across page and block templates.

### Headings

> Editors shouldn't have to manually set heading levels. The templates should be able to infer the correct heading tag from the page's content hierarchy. We do that by incrementing a template variable every time we go down the blocks nesting.

Reviewer: 👏

Consider the following page:

- Notification - no heading here
- Into block
    - Heading (`<h1>`)
    - Some content
- Section
    - Section heading (`<h2>`)
    - Section content
        - Cards list
            - Card
                - Card heading (`<h3>`)
                - Card content
            - Other cards...
- Banner
    - Banner heading (`<h2>`)
    - Banner content

Each of these blocks have headings. To determine which tag to use, they need to know their position on the page. The first block on the page that has a heading needs an `h1`. Every other block in the first level needs an `h2`. The children of these blocks need to be one level lower: `h3`. If they had children, those should get an `h4`, and so on...

Here's how we implement that.

`page.html`

> ```html

Reviewer 2: ```suggestion
``` jina
```

🐼 Pedantry: Code highlighting works better with `jinja`.

{% block content %}
<div class="my-page-class">
  {% set ns = namespace(headings=0) %}
  {% set block_level = 1 %}
  {% for block in page.content %}
    {% set block_index = loop.index %}
    {% set block_type = block.block_type %}
    {% set block_position = "block-" ~ block_index ~ "-" ~ block_type %}

    {% if (block.value.heading and block.value.heading.heading_text) or block.value.headline %}
      {% set block_level = 1 if ns.headings == 0 else 2 %}
      {% set ns.headings = ns.headings + 1 %}
    {% endif %}

    {% include_block block %}
  {% endfor %}
</div>
{% endblock content %}
```

The page creates a `headings` variable to count the blocks that have headings. Note that we can't just use the loop index because some blocks don't have headings, such as notifications. This count is used to set the `block_level` variable. The first block gets `block_level = 1` and all the following ones get `block_level = 2`.

When the heading block template is rendered, it will use this variable to determine the heading tag.

`blocks/heading.html`

> ```html

Reviewer 2: ```suggestion
``` jinja
```

- `data-cta-text` - This is the human readable attribute. It should be composed of the text of the nearest heading and the button's text. For example: "Cool new feature launched in Firefox - Learn more".

Note the `block_position` variable on the page template above. This variable should always be set at the page level and then incremented by each block down the blocks' tree.

The block position should follow this pattern:

- Levels are separated by `.`
- Loops are indicated by indexes
- Block types identify the blocks

`block-{index}-{block_type}.{child_block_type}.item-{index}-{grandchild_block_type}.button-1`

To determine the `data-cta-text`, every block that has a heading, must set the `block_text` variable.

Here's a full example for the `block_text` and `block_position` usage.

Consider this hierarchy:

- Section
    - Section heading
    - Section content
        - Cards list
            - Card
                - Card heading
                - Card content
                - Card button
            - More cards...
    - Call to action (button)

The section block's html would look like this:

`section.html`

> ```html

Reviewer 2: ```suggestion
``` jinja
```

{% set block_text = value.heading.heading_text|richtext|remove_tags %}

<div class="section">
  <div class="section-heading">
    {% include_block value.heading %}
  </div>

  <div class="section-content">
    {% for block in value.content %}
      {% set block_position = block_position ~ ".item-" ~ loop.index ~ "-" ~ block.block_type %}
      {% with block_level = block_level + 1 if block_level else 2 %}
        {% include_block block %}
      {% endwith %}
    {% endfor %}
  </div>

  {% if value.cta %}
    <div class="section-cta">
      {% for button in value.cta %}
        {% set block_position = block_position ~ ".cta-" ~ loop.index %}
        {% include_block button %}
      {% endfor %}
    </div>
  {% endif %}
</div>
```

In this example, both the Call to Action and the cards's button use the a button block.

Here's what the button template looks like:

`button.html`

> ```html

Reviewer 2: ```suggestion
``` jinja
```

{% set block_text = block_text ~ " - " ~ value.label %}

<button
  href="{{ value.url }}"
  data-cta-text="{{ block_text }}"
  data-cta-position="{{ block_position }}"
  data-cta-uid="{{ value.settings.analytics_id }}"
>
  {{ value.label }}
</button>
```

The section set the `block_text` variable with the content from it's heading. Note that we're using a custom `remove_tags` template tag to preserve only the text, without any HTML tags.

> When the button is rendered in the CTA block, it will use the existing `block_text` template variable and append it's own label.  The `block_position` is set by the parent block. The result for the CTA button will be `data-cta-position="block-1-section.cta-1"` and `data-cta-text="Section Heading - Button label"`.

Reviewer: ```suggestion
When the button is rendered in the CTA block, it will use the existing `block_text` template variable and append its own label.  The `block_position` is set by the parent block. The result for the CTA button will be `data-cta-position="block-1-section.cta-1"` and `data-cta-text="Section Heading - Button label"`.
```

Now let's look at the cards from the example. They are in a lower level in relation to the section.

`cards-list.html`

> ```html

Reviewer 2: ```suggestion
``` jinja
```

{% set block_level = block_level or 1 %}

<div class="cards-list">
  {% for card in value.cards %}
    {% set card_index = loop.index %}
    {% set block_position = block_position ~ ".card-" ~ card_index %}

    {% include_block card %}
  {% endfor %}
</div>
```

`card.html`

> ```html

Reviewer 2: ```suggestion
``` jinja
```

        try:
            content_type = ContentType.objects.get(app_label=app_label, model=model_name)
        except ContentType.DoesNotExist:
            continue

        for revision in Revision.objects.filter(base_content_type=content_type).iterator():
            modified = False
            for field_name in field_names:
                raw_json = revision.content.get(field_name)
                if not raw_json:
                    continue
                try:
                    field_data = json.loads(raw_json)
                    if update_banner_themes(field_data):
                        revision.content[field_name] = json.dumps(field_data)
                        modified = True
                except (json.JSONDecodeError, TypeError):
                    pass
            if modified:
                revision.save(update_fields=["content"])

class Migration(migrations.Migration):
    dependencies = [
        ("cms", "XXXX_previous_migration"),
    ]

    operations = [
        migrations.RunPython(convert_pages, reverse_code=migrations.RunPython.noop),
        migrations.RunPython(convert_revisions, reverse_code=migrations.RunPython.noop),
    ]
```

> Note that the migration finds the block's data by the key used in the Stream Field's definition. Pages and blocks need to be consistent with the keys they use for the same blocks.

Reviewer: **non-blocking** Maybe also note what NOT to do in migrations (i.e. avoid Alter & Drop together?) Or point to the general migration docs for more information? https://mozmeao.github.io/platform-docs/operations/migrations/

Author: This type of migration doesn't touch the database structure; it only alters data.
